catalog/db_batch_entry_read.go

```diff
@@ -0,0 +1,166 @@
+package catalog
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/treeverse/lakefs/db"
+)
+
+const (
+	MaxReadQueue        = 10
+	ReadEntryTimeout    = time.Second * 15
+	ScanTimeout         = time.Microsecond * 500
+	waitTimeout         = time.Microsecond * 1000
+	MaxEntriesInRequest = 64
+	ReadersNum          = 8
+)
+
+type pathRequest struct {
+	path      string
+	replyChan chan readResponse
+}
+
+type readRequest struct {
+	bufKey  bufferingKey
+	pathReq pathRequest
+}
+type readResponse struct {
+	entry *Entry
+	err   error
+}
+
+type readBatch struct {
+	startTime time.Time
+	pathList  []pathRequest
+}
+
+type bufferingKey struct {
+	repository string
+	ref        Ref
+}
+
+type batchReadMessage struct {
+	key   bufferingKey
+	batch []pathRequest
+}
+
+func (c *cataloger) initBatchEntryReader() {
+	go c.readOrchestrator()
+	c.wg.Add(1)
+	for i := 0; i < ReadersNum; i++ {
+		go c.readEntriesBatch()
+		c.wg.Add(1)
+	}
+}
+
+func (c *cataloger) dbBatchEntryRead(repository, path string, ref Ref) (*Entry, error) {
+	replyChan := make(chan readResponse, 1) // channel closed by readEntriesBatch
+	request := &readRequest{
+		bufferingKey{repository, ref},
+		pathRequest{path, replyChan},
+	}
+	c.readEntryRequestChan <- request
+	select {
+	case response := <-replyChan:
+		return response.entry, response.err
+	case <-time.After(ReadEntryTimeout):
+		return nil, ErrReadEntryTimeout
+	}
+}
+
+func (c *cataloger) readOrchestrator() {
+	defer c.wg.Done()
+	bufferingMap := make(map[bufferingKey]*readBatch)
+	timer := time.NewTimer(0)
+	for {
+		if len(bufferingMap) > 0 {
+			timer.Reset(ScanTimeout)
```

Comment:
This is a fairly busy loop: it keeps going around doing nothing even when it has almost nothing to do.  So (with the above parameters) on a lightly-loaded system (say around 1000rps) it will wake up and loop through 2000 times per second.  That can be very wasteful given that we know exactly when the next timeout will be.
Similarly, on a loaded system it keeps on running through all the branches.  E.g. suppose we have 20_000 rps on master and 100 rps on each of 100 other branches (total 30_000 rps across all 101 branches).  Then 30_000 times a second we wake up and go over 101 branches (so 3_000_000 checks a second) but we only flush 2000 times per second for master and 10 times per second for each of the 100 other branches, for a total of 2100 times a second.  So this is inefficient in CPU.

Reply:
I see your point. will change to scanning only when a timer happens